### ISSUE TITLE
Delayed DAO update flush causes unique constraint violation

### ISSUE THREAD

**Issue by kiliangrashoff** [JetBrains/Exposed#717]
When the following steps occur, a constraint violation exception is thrown:

1. DAO entity _X_ is created with value _a_ for a column with a unique index
2. A lazy reference to _X_ is read
3. Value _a_ for _X_ is updated to value _b_
4. A second entity _Y_ is created with value _a_

This sequence of steps should not cause a unique constraint violation, but it does because the update is delayed until after the second insert. Without step 2, exposed would merge step 1 and 3 together into one insert statement. That does not happen in this case, since reading the lazy reference flushes the insert for step 1.

Exposed version: 0.18.1
Database: PostgreSQL 11

### Expected behaviour
The update for step 3 should be executed before the insert for step 4. If the DAO does not respect unique constraints in this case it should be documented, since this is unexpected behaviour that breaks the DAO abstraction.

### Code to reproduce
```
import org.jetbrains.exposed.dao.EntityID
import org.jetbrains.exposed.dao.IntEntity
import org.jetbrains.exposed.dao.IntEntityClass
import org.jetbrains.exposed.dao.IntIdTable
import org.jetbrains.exposed.sql.Database
import org.jetbrains.exposed.sql.SchemaUtils
import org.jetbrains.exposed.sql.StdOutSqlLogger
import org.jetbrains.exposed.sql.addLogger
import org.jetbrains.exposed.sql.deleteAll
import org.jetbrains.exposed.sql.transactions.transaction
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test

object Parents : IntIdTable("parents")

class Parent(id: EntityID<Int>) : IntEntity(id) {
	companion object : IntEntityClass<Parent>(Parents)

	val children by Child referrersOn Children.parent
}

object Children : IntIdTable("children") {
	val number = integer("number").uniqueIndex()
	val parent = reference("parents", Parents)
}

class Child(id: EntityID<Int>) : IntEntity(id) {
	companion object : IntEntityClass<Child>(Children)

	var number by Children.number
	var parent by Parent referencedOn Children.parent
}

class ExposedTest {
	init {
		Database.connect("<postgres url>", "<driver>", "<user>", "<password>")
		transaction {
			SchemaUtils.create(Parents, Children)
		}
	}

	@Test
	fun `Unique constraints should not be violated by delayed update`() {
		transaction {
			addLogger(StdOutSqlLogger)
			val parentEntity = Parent.new (1) {}
			val example = Child.new {
				number = 1
				parent = parentEntity
			}
			// This call prevents folding the update into the previous insert.
			parentEntity.children.toList()
			example.number = 2
			Child.new {
				number = 1
				parent = parentEntity
			}
		}
	}

	@AfterEach
	fun afterTest() {
		transaction {
			Children.deleteAll()
			Parents.deleteAll()
		}
	}
}
```

### Output
```
SQL: INSERT INTO parents (id) VALUES (1)
SQL: INSERT INTO children ("number", parents) VALUES (1, 1)
SQL: SELECT children.id, children."number", children.parents FROM children WHERE children.parents = 1
SQL: INSERT INTO children ("number", parents) VALUES (1, 1)
```
The update never executes due to the unique constraint. If we remove the constraint, we can see that it is executed after the last insert:
```
SQL: UPDATE children SET "number"=2 WHERE id = 1
```
### Error
```
WARN Exposed - Transaction attempt #0 failed: java.sql.BatchUpdateException: Batch entry 0 INSERT INTO children ("number", parents) VALUES (1, 1)
RETURNING * was aborted: ERROR: duplicate key value violates unique constraint "children_number_unique"
  Detail: Key (number)=(1) already exists.  Call getNextException to see other errors in the batch.. Statement(s): INSERT INTO children ("number", parents) VALUES (?, ?)
org.jetbrains.exposed.exceptions.ExposedSQLException: java.sql.BatchUpdateException: Batch entry 0 INSERT INTO children ("number", parents) VALUES (1, 1)
RETURNING * was aborted: ERROR: duplicate key value violates unique constraint "children_number_unique"
  Detail: Key (number)=(1) already exists.  Call getNextException to see other errors in the batch.
SQL: [INSERT INTO children ("number", parents) VALUES (?, ?)]
	at org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:63)
<truncated>
Caused by: java.sql.BatchUpdateException: Batch entry 0 INSERT INTO children ("number", parents) VALUES (1, 1)
RETURNING * was aborted: ERROR: duplicate key value violates unique constraint "children_number_unique"
  Detail: Key (number)=(1) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:148)
<truncated>
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "children_number_unique"
  Detail: Key (number)=(1) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2468)
<truncated>
```

**Comment by YegorBabarykin** [comment]
I had the same bug when I tried to upgrade version of library from 0.17.7 to 0.18.1

**Comment by Tapac** [comment]
Fixed in master, will be available in 0.19.2
